internal/work/results/guarantee.go

```diff
@@ -0,0 +1,164 @@
+func (gp *GuaranteeManager) ProcessWorkPackageGuarantee(
+	wp work.Package,
+	coreIndex uint16,
+	validatorIndex uint16,
+	validatorPrivateKey ed25519.PrivateKey,
+	validatorPublicKeys map[uint16]ed25519.PublicKey,
+	authPool state.CoreAuthorizersPool, // Authorization pool from recent chain state
+) (*block.Guarantee, error) {
+	if validatorPrivateKey == nil {
+		return nil, errors.New("validator key cannot be nil")
+	}
+	if validatorPublicKeys == nil {
+		return nil, errors.New("public keys cannot be nil")
+	}
+
+	// Validate work package authorization
+	if err := gp.validateWorkPackageAuthorization(wp, coreIndex, authPool); err != nil {
+		return nil, fmt.Errorf("authorization validation failed: %w", err)
+	}
+
+	// Generate and validate work report.
+	workReport, err := gp.computation.EvaluateWorkPackage(wp, coreIndex)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate work-report: %w", err)
+	}
+
+	// Validate work report output size
+	if !workReport.OutputSizeIsValid() {
+		return nil, errors.New("work report output size exceeds limit")
+	}
+
+	// Generate guarantee with validator signature
+	guarantee, err := gp.GenerateGuarantee(workReport, validatorIndex, validatorPrivateKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate guarantee: %w", err)
+	}
+
+	// Validate the guarantee
+	if err := ValidateGuarantee(guarantee, validatorPublicKeys); err != nil {
+		return nil, fmt.Errorf("guarantee validation failed: %w", err)
+	}
+
+	return guarantee, nil
+}
+
+// validateWorkPackageAuthorization checks if the work package is authorized
+// according to the current authorization pool
+func (gp *GuaranteeManager) validateWorkPackageAuthorization(
+	wp work.Package,
+	coreIndex uint16,
+	authPool state.CoreAuthorizersPool,
+) error {
+	// Get authorizers for this core
+	authorizers := authPool[coreIndex]
+
+	authorized := false
+	for _, authHash := range authorizers {
+		if authHash == wp.AuthCodeHash {
+			authorized = true
+			break
+		}
+	}
+
+	if !authorized {
+		return errors.New("work package not authorized for this core")
+	}
+
+	return nil
+}
+
+// GenerateGuarantee creates a signed guarantee for a Work Report
+func (gp *GuaranteeManager) GenerateGuarantee(
+	workReport *block.WorkReport,
+	validatorIndex uint16,
+	validatorPrivateKey ed25519.PrivateKey,
+) (*block.Guarantee, error) {
+	if workReport == nil {
+		return nil, errors.New("work report cannot be nil")
+	}
+
+	reportHash, err := workReport.Hash()
+	if err != nil {
+		return nil, fmt.Errorf("failed to compute work-report hash: %w", err)
+	}
+
+	signature := ed25519.Sign(validatorPrivateKey, reportHash[:])
+
+	guarantee := &block.Guarantee{
+		WorkReport: *workReport,
+		Timeslot:   jamtime.CurrentTimeslot(),
+		Credentials: []block.CredentialSignature{
+			{
+				ValidatorIndex: validatorIndex,
+				Signature:      ([crypto.Ed25519SignatureSize]byte)(signature[:crypto.Ed25519SignatureSize]),
```

Comment:
Why do we need to limit the signature size here, wouldn't the signature be a fixed size anyway and what happens if it's bigger and gets cut off?

Reply:
Also we already seem to have `[crypto.Ed25519SignatureSize]byte` defined here: https://github.com/eigerco/strawberry/blob/main/internal/crypto/keys.go#L9